torch/csrc/jit/frontend/sugared_value.h

```diff
@@ -458,15 +458,37 @@ struct MethodValue : public SugaredValue {
     std::vector<const FunctionSchema*> schemas;
     for (const std::string& method_name : method_names_) {
       if (auto class_type = self_->type()->cast<ClassType>()) {
-        Function& method = class_type->getMethod(method_name);
-        try {
-          method.ensure_defined();
-        } catch (const RecursiveMethodCallError&) {
-          throw ErrorReport(loc)
-              << " method '" << method.name() << "' is called recursively. "
-              << "Recursive calls are not supported";
+        Function* method = class_type->findMethod(method_name);
+        if (auto overloaded_method =
+                dynamic_cast<OverloadedFunction*>(method)) {
+          auto overloaded_schemas =
+              overloaded_method->loadPossibleSchemas(class_type, loc);
+          auto matched = matchSchemas(
+              overloaded_schemas, loc, *f.graph(), argsWithSelf, kwargs);
+
+          // When we import the graph back (serialization), we will sometimes
+          // receive mangled names. Therefore we need to make sure we don't add
+          // extra __[number] tag again.
+          auto mangled_name = method_name;
```

Comment:
I think this is a sign of something we need to clean up. When we serialize an invoked overloaded method, we should be serializing the qualified name of a particular overload. That particular overload should be an actual method on the class which gets serialized correctly, and the OverloadedFunction should have a reference to the list of methods which it resolves to.

Reply:
+1

Reply:
I changed it so that we only serialize only the qualified name. Could you guys take a look? It makes the serialization code bit more complex but at least we don't have to deal with redundant mangling anymore